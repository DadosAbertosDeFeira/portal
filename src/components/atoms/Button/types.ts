import type { ComponentPropsWithRef } from "react";

export type ButtonVariants = "outline" | "condensed" | "default";

export type ButtonProps = ComponentPropsWithRef<"button"> & {
  variant: ButtonVariants;
};
