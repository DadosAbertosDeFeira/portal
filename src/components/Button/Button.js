import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import styles from './Button.module.scss';

const noop = () => {};

const Button = ({
  onClick,
  secondary,
  className,
  rounded,
  small,
  medium,
  large,
  children,
  icon,
}) => (
  <button
    type="button"
    className={classnames(
      styles.buttonDefault,
      className && className,
      secondary && styles.secondary,
      rounded && styles.rounded,
      small && styles.small,
      medium && styles.medium,
      large && styles.large
    )}
    onClick={onClick}
  >
    {icon && <icon />}
    {children}
  </button>
);

Button.propTypes = {
  onClick: PropTypes.func,
  className: PropTypes.string,
  rounded: PropTypes.bool,
  small: PropTypes.bool,
  medium: PropTypes.bool,
  large: PropTypes.bool,
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  icon: PropTypes.node,
  secondary: PropTypes.bool,
};

Button.defaultProps = {
  onClick: noop,
  className: '',
  rounded: false,
  small: false,
  medium: false,
  large: false,
  children: '',
  icon: null,
  secondary: false,
};

export default Button;
