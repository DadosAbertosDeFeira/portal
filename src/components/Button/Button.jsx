/* eslint-disable react/prop-types */
import React from "react";
import classNames from "classnames";

export const Button = ({ className, children, variant, ...props }) => {
  return (
    <button
      type="button"
      {...props}
      className={classNames(
        "font-semibold p-3 rounded transition ease-in-out flex justify-center items-center",
        variant === "outline" && "border border-primary-dark text-primary-dark",
        variant === "condensed" && "bg-primary-dark text-white",
        className
      )}
    >
      {children}
    </button>
  );
};

export default Button;
