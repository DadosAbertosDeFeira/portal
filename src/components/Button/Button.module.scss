@import "../../styles/variables";

.buttonDefault {
  cursor: pointer;
  border-radius: 1px;
  padding: 8px 16px;
  background-color: $color-primary;
  color: $color-white;
  border: none;
  font-size: 14px;
  line-height: 1.75rem;
  outline: none;
}

.buttonDefault:active {
  background-color: $color-primary--dark;
}

.secondary {
  background-color: $color-secondary;
  color: $color-primary--dark;
}

.secondary:active {
  background-color: $color-secondary--dark;
}

.rounded {
  border-radius: 56px;
}

/* Sizes */
.small {
  font-size: 12px;
  padding: 4px 8px;
}

.large {
  font-size: 16px;
  padding: 12px 24px;
}

.typeText {
  background: none;
  color: $color-primary;
}

.typeText:active {
  background: none;
  color: $color-primary--dark;
}

.typeText .secondary {
  color: $color-secondary;
}

.typeText .secondary:active {
  background: none;
  color: $color-secondary--dark;
}
