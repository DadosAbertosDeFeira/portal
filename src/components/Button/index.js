import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './styles.module.scss';

export default function Button({ outline, onClick, children, ...rest }) {
  return (
    <button
      type="button"
      className={classnames(
        styles.button,
        outline ? styles.outline : styles.full
      )}
      onClick={onClick}
      {...rest}
    >
      {children}
    </button>
  );
}

Button.defaultProps = {
  outline: false,
  onClick: () => {},
};

Button.propTypes = {
  outline: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  onClick: PropTypes.func,
};
