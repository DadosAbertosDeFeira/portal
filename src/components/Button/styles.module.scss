@import '../../styles/variables.scss';

.button {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;

  width: 290px;
  height: 53px;
  border-radius: 4px;
  cursor: pointer;
}

#outline {
  border: 1px solid $color-primary--dark;
  background: $color-white;
  color: $color-primary--dark;
  font-weight: 600;
}

#full {
  border: 0;
  background: $color-primary--dark;
  color: $color-white;
  font-weight: 600;
}
