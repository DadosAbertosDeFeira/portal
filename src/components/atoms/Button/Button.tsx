import { forwardRef } from "react";
import { twMerge } from "tailwind-merge";

import type { ButtonProps, ButtonVariants } from "./types";

export const buttonVariants: Record<ButtonVariants, string> = {
  condensed: "bg-blue-700 text-white",
  "condensed-rounded": "bg-blue-700 text-white rounded-full",
  default: "text-blue-700 active:bg-blue-100/10",
  outline: "border border-blue-700 text-blue-700",
};

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    { children, variant = "default", type = "button", className, ...props },
    ref
  ) {
    return (
      <button
        className={twMerge(
          "font-semibold p-3 rounded transition ease-in-out flex justify-center items-center",
          buttonVariants[variant],
          className
        )}
        type={type}
        {...props}
        ref={ref}
      >
        {children}
      </button>
    );
  }
);
