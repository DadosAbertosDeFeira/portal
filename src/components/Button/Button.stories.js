/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import Button from './Button';

export const ButtonTemplate = (args) => <Button {...args} />;
ButtonTemplate.args = { children: 'Button' };
ButtonTemplate.storyName = 'Default';

export default {
  title: 'Components/Button',
  component: Button,
};
