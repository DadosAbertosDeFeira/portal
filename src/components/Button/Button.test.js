/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { FiActivity } from 'react-icons/fi';

import Button from './Button';

describe('<Button />', () => {
  it('renders correctly by default', () => {
    const { asFragment } = render(<Button>Dados Abertos</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly as secondary', () => {
    const { asFragment } = render(<Button secondary>Dados Abertos</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with text type', () => {
    const { asFragment } = render(<Button typeText>Dados Abertos</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly rounded', () => {
    const { asFragment } = render(<Button rounded>Dados Abertos</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly small size', () => {
    const { asFragment } = render(<Button small>Dados Abertos</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly medium size', () => {
    const { asFragment } = render(<Button medium>Dados Abertos</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly large size', () => {
    const { asFragment } = render(<Button large>Dados Abertos</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with an icon and text', () => {
    const { asFragment } = render(
      <Button icon={FiActivity}>Dados Abertos</Button>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correclty just with an icon', () => {
    const { asFragment } = render(<Button icon={FiActivity} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly using html attributes', () => {
    const { asFragment } = render(
      <Button name="dados-button">Dados Abertos</Button>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('events', () => {
    it('handle the when user clicks', () => {
      const onClick = jest.fn();
      const { getByTestId } = render(
        <Button onClick={onClick}>Dados Abertos</Button>
      );
      fireEvent.click(getByTestId('button'));

      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });
});
