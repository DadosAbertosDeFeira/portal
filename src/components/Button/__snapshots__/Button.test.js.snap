// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly outline button 1`] = `
<DocumentFragment>
  <button
    class="button outline new-class"
    type="button"
  >
    Click Me!
  </button>
</DocumentFragment>
`;

exports[`<Button /> renders correctly with default properties 1`] = `
<DocumentFragment>
  <button
    class="button full"
    type="button"
  >
    Click Me!
  </button>
</DocumentFragment>
`;
